hadoop-hdds/docs/content/design/storage-class.md

```diff
@@ -19,10 +19,331 @@ author: Marton Ele
+
+Keys (and containers) can be categorized by storage-class which determines the replication scheme.
+
+## [ADMIN/USER] Flexible administrations
+
+As it's mentioned above, today it's hard to configure the details of the replications for key/bucket level. The only thing what we can define is the replication type for open containers (RATIS/THREE or RATIS/ONE) which determines the later lifecycle of the keys/containers.
+
+Any specific replication configuration can be configured only on cluster level and not on key level.
+
+A storage-class can define all the parameters for the spefific containers/keys:
+
+As an example this could be a storage-class definitions:
+
+```
+name: STANDARD
+states:
+    - name: open
+      replicationType: RATIS
+      repliationFactor: THREE
+    - name: closed
+      replicationType: COPY
+      repliationFactor: TWO
+      rackPolicy: different
+      transitions:
+        - target: ec
+          trigger:
+             ratio: 90%
+             used: 30d
+    - name: ec
+      codec: Reed-Solomon
+      scheme:
+         data: 6
+         parity: 3
+```
+
+This defines a replication scheme where only two replicas are enough from closed containers, and container will be erasure encoded under the hood if the 90% of the content is not used in the last 30 days.
+
+Please note that:
+
+ * All the low-level details of the replication rules can be configured here by the administrators
+ * Configuration is not global and not cluster-level, one can have different configuration for different storage-classes (which means for different keys/containers)
+ * Users dont' need to face with these details as they can use the storage-class (or just use the pre-created buckets and use default storage-class) abstraction
+
+## [DEV] Give flexibility to the developers
+
+Storage-class abstraction provides an easy way to plug in newer replication schemes. New type of replications (like EC) can be supported easily as the system will be prepared to allocate different type of containers.
+
+## [ADMIN] Better upgrade support
+
+Let's imagine that a new type of Open container replication is introduced (`RATIS-STREAM/THREE` instead of `RATIS/THREE`). If storage-classes are stored with the keys and containers instead of the direct replication rules we can:
+
+ 1. Easily change the replicaiton method of existing buckets/keys
+ 2. Turn on experimental features for specific buckets
+
+
+## [ADMIN] Change the cluster-wide replication
+
+An admin may decide to set a custom policy for an entire cluster.
+
+```
+ozone sh prefix setClass --storage-class=EC_6_3 /
+```
+
+# Unsupported use cases
+
+The following use cases are specifically unsupported.
+
+## [USER] Change the replication policy for a pre-existing key
+
+Changing the replication policy for a pre-existing key will require data movement and reauthoring containers and hence it is unsupported.
+
+## [USER] Defining storage-classes using Hadoop Compatible File System interface
```

Comment:
Hadoop apps use setReplication(), can we support that in storage class?

Reply:
Good point. We can support it with mapping numbers to storage classes:

```
setReplication(1) --> StorageClass=REDUCED (can be STANALONE/ONE->CLOSED/ONE or different)
setReplication(3) --> StorageClass=STANDARD (can be RATIS/THREE or EC or anything else...)
```